### PR TITLE
Don't update unchanged session so often - performance issue

### PR DESCRIPTION
**Pull request by ianshward**
This pull request changes how #set() works, whereby if the current session in couch is the same as the session which is about to be `put` to couch, except for timestamps, like lastAccess, expires, and originalMaxAge, and less than 1 minute has passed since the last `put` for this session, then, don't put the session.

Performance wise, the `put` operation is very slow.  I've benchmarked the performance impact on a stack of two servers with couchdb replication running for the session database, and observe the current situation (`put` no matter what) at **25 r/s** compared to `put` only if something significant in the session changed or if > 1m has passed since this session was last put at **105 r/s** which is a major improvement.

The consequence of lots of puts at the second or sub-second level leads to couchdb becoming slow, dealing with the flurry of conflicts and replication.  Once you have maybe 20 or 40 concurrent users surfing around on your web app this becomes a real bottleneck / scalability breaking point.

I've also written a test that tests a session really is not `put` if nothing significant changes in under a minute, and that the session really is changed if significant data changes and is then attempted to be `set`.

### CODE REVIEW

**Comment by tdebarochez** [comment]
That's a really great idea ! But I've two questions :
- Are you sure that the locks object is necessary ? Why don't you just compare lastAccess timestamps (if existing) ?
- Does the minute could be easily configurable ?

**Comment by ianshward** [comment]
Good point about the locks object and lastAccess, I was so fixed on the problem of getting rid of those properties in order to do the comparison that I did not see their usefulness for this situation :)  I also made the 60000ms a setting - `setThrottle`

File: lib/connect-couchdb.js [hunk 1]
```diff
@@ -115,9 +116,43 @@ module.exports = function(connect) {
...
+                var _lastAccess = sess.lastAccess;
+                var _originalMaxAge = sess.cookie.originalMaxAge;
+                if (sess.cookie._expires) {
+                    var _expires = sess.cookie._expires;
+                    sess.cookie._expires = null;
+                }
+                if (doc.sess.cookie.expires) {
+                    var _expiresDoc = doc.sess.cookie.expires;
+                    doc.sess.cookie.expires = null;
+                }
```
Comment (tdebarochez) [inline lib/connect-couchdb.js:129]: What's the utility about `doc.sess.cookie.expires` and `_expiresDoc` variables ? You never use there...

File: lib/connect-couchdb.js [hunk 2]
```diff
@@ -115,9 +116,43 @@ module.exports = function(connect) {
...
+                    this.db.put(doc, fn);
+                } else {
+                    // Otherwise, only save if > 1s has elaspsed since last save.
+                    if (accessGap > this.setThrottle) {
+                        sess.lastAccess = _lastAccess;
+                        if (_expires) sess.cookie._expires = _expires;
+                        sess.cookie.originalMaxAge = _originalMaxAge;
+                        doc.expires = expires;
+                        doc.sess = sess;
+                        this.db.put(doc, fn);
```
Comment (tdebarochez) [inline lib/connect-couchdb.js]: Is this code block the same as above ?
Couldn't you write :

```
if (JSON.stringify(doc.sess) !== JSON.stringify(sess)
    || accessGap > this.setThrottle) {
    sess.lastAccess = _lastAccess;
    if (_expires) sess.cookie._expires = _expires;
    sess.cookie.originalMaxAge = _originalMaxAge;
    doc.expires = expires;
    doc.sess = sess;
    this.db.put(doc, fn);
} 
else {
    return fn();
}
```

I hate code duplication :)

**Comment by tdebarochez** [comment]
Do not you think this improvement should be extends to every connect's session system (or should be an option of it) ?

**Comment by ianshward** [comment]
I made the fixes you pointed out above.  That variable was just left over from a previous refactor.  What's implemented in connect seems pretty minimal, not sure if it would make sense to add this logic at that level.  For example this isn't very helpful for the memory store.  Not sure what other stores would make sense to use this.

**Comment by tdebarochez** [comment]
Thanks a lot !

**Comment by ianshward** [comment]
Thanks Thomas.  Are you planning to make a new release soon?

**Comment by tdebarochez** [comment]
Only if someone need it.

**Comment by ianshward** [comment]
We could use one, we'd like to roll out with the latest changes.

**Comment by tdebarochez** [comment]
Done.

**Comment by ianshward** [comment]
Thanks!

**Pull request merged**